{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#install.packages(c(\"ape\",\"ggplot2\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "executionInfo": {
     "elapsed": 34511,
     "status": "ok",
     "timestamp": 1623036673757,
     "user": {
      "displayName": "Tsu-Wang Sun",
      "photoUrl": "",
      "userId": "15932095404451727917"
     },
     "user_tz": -480
    },
    "id": "J5qIi0EhbSgB",
    "outputId": "358b05ac-5c10-4b5c-f562-4f8ac6bcbf15"
   },
   "outputs": [],
   "source": [
    "library(ape)\n",
    "library(ggplot2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 52
    },
    "executionInfo": {
     "elapsed": 5303,
     "status": "ok",
     "timestamp": 1623036792554,
     "user": {
      "displayName": "Tsu-Wang Sun",
      "photoUrl": "",
      "userId": "15932095404451727917"
     },
     "user_tz": -480
    },
    "id": "spe43ir7xDje",
    "outputId": "66e5690b-296d-4f90-8369-519fbf3910e3"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1234567"
     ]
    },
    {
     "data": {
      "text/html": [
       "<strong>png:</strong> 2"
      ],
      "text/latex": [
       "\\textbf{png:} 2"
      ],
      "text/markdown": [
       "**png:** 2"
      ],
      "text/plain": [
       "png \n",
       "  2 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "vdir=\"\"#\"../from_list_01/\"\n",
    "tdir=\"\"#../fig1/concatenated_tree/\n",
    "md=\"0807\"\n",
    "###Host diversity\n",
    "nspl=read.csv(paste0(vdir,\"viruslist-single-pan\",md,\"_plus2.tsv\"),header=T,sep='\\t')\n",
    "colnames(nspl)[which(colnames(nspl)=='Lineage')]='virTree'\n",
    "colnames(nspl)[which(colnames(nspl)=='Host_lineage')]='EukTree'\n",
    "\n",
    "levels(nspl$virTree)=sort(as.character(sub(\" \",\"_\",sub(\" $\",\"\",unique(nspl$virTree)))))\n",
    "euktax=read.csv(paste0(\"euktax.tsv\"),header=T,sep='\\t')\n",
    "#nspl=nspl[-which(nspl$EukTree==\"\"),]\n",
    "nspl[which(nspl$Family=='Mimiviridae-P'),'Family']='Mimiviridae'\n",
    "nspl[which(nspl$Family=='Pandoraviridae'),'Family']='Phycodnaviridae'\n",
    "nspl[which(nspl$Family=='Ascoviridae'),'Family']='Iridoviridae'\n",
    "\n",
    "eutx=NULL\n",
    "for(i in 1:nrow(euktax)){\n",
    "  eutx[grep(euktax$EukTree[i],nspl$EukTree)]=as.character(euktax$hostLineage[i])\n",
    "}\n",
    "cspl=cbind(nspl,eutx)\n",
    "\n",
    "# hdiv=aggregate(.~Family,cspl,function(x) c(mean=mean(x),sd=sd(x),len=length(unique(x))))\n",
    "# #write.table(unique(spl$EukTree),paste0(vdir,\"visualization/pair_compare/\",\"euktax\"),quote=F,row.names=F,col.names=F)\n",
    "# plot(hdiv$eutx[,'len'],hdiv$Length[,'sd'])\n",
    "# plot(hdiv$eutx[,'len'],hdiv$gf.unique.fm[,'mean'])\n",
    "# \n",
    "# ggplot(hdiv,aes(x=hdiv$eutx[,'len'],y=hdiv$aanum[,'sd']))+\n",
    "#   geom_point()+theme_classic()+ geom_text(aes(label=hdiv$Family)) \n",
    "panlin=read.csv(paste0(vdir,\"pangenome\",md,\"lineage.tsv\"),header=T,sep='\\t')\n",
    "\n",
    "# entropy\n",
    "entt=function(ps){\n",
    "  ps=ps[which(ps>0)]\n",
    "  ans=0.000001\n",
    "  for(i in 1:length(ps)){ \n",
    "    ans=ans+(ps[i]*log(ps[i]))\n",
    "  }\n",
    "  return(-ans)\n",
    "}\n",
    "\n",
    "\n",
    "# core gene seq div\n",
    "library(ape)\n",
    "tre1=read.tree(file=paste0(tdir,\"iq2-allmaf_og.fasta.treefile\"))\n",
    "kalldist=NULL\n",
    "for(i in 1:nrow(cspl)){\n",
    "  kalldist[i]=cophenetic.phylo(tre1)['P49',cspl[i,'Abre']]\n",
    "}\n",
    "cspl=cbind(cspl,kalldist)\n",
    "\n",
    "\n",
    "#abl=cspl[which(cspl$Family==hdiv2[i,'family']),'Abre']\n",
    "\n",
    "corediv=function(abl){\n",
    "  alldist=0\n",
    "  for(j in abl){\n",
    "    alldist=alldist+mean(cophenetic.phylo(tre1)[j,as.character(abl[-which(abl==j)])])\n",
    "  }\n",
    "  return(alldist/length(abl))\n",
    "}\n",
    "\n",
    "\n",
    "lincorediv=function(i){\n",
    "  lineage_pool=unique(cspl$virTree[which(cspl$Family==unique(cspl$Family)[i])])\n",
    "  lins2lins=0\n",
    "  #abl=cspl$Abre[which(cspl$Family==unique(cspl$Family)[i])]\n",
    "  for(v in lineage_pool){\n",
    "    lin2lins=0\n",
    "    lineage_pool=lineage_pool[-which(lineage_pool==v)]#c(used,as.character(cspl[which(cspl$virTree==v),'Abre'])\n",
    "    if(length(lineage_pool)==0){break;}\n",
    "    for(j in cspl[which(cspl$virTree==v),'Abre']){\n",
    "      gm2lins=0\n",
    "      for(k in lineage_pool){\n",
    "        gm2lins=gm2lins+mean(cophenetic.phylo(tre1)[as.character(j),as.character(cspl[which(cspl$virTree==k),'Abre'])])\n",
    "      }\n",
    "      lin2lins=lin2lins+gm2lins/length(lineage_pool)\n",
    "    }\n",
    "    #lins2lins=lins2lins+lin2lins/length(cspl[which(cspl$virTree==v),'Abre'])\n",
    "    lins2lins=c(lins2lins,lin2lins/length(cspl[which(cspl$virTree==v),'Abre']))\n",
    "  }\n",
    "  #lins2lins/(length(unique(cspl$virTree[which(cspl$Family==unique(cspl$Family)[i])]))-1)\n",
    "  return(lins2lins[-1])\n",
    "}\n",
    "\n",
    "\n",
    "\n",
    "#lineage mean\n",
    "linlin=function(i,colm){\n",
    "  ans=NULL\n",
    "  for(v in unique(cspl$virTree[which(cspl$Family==unique(cspl$Family)[i])])){\n",
    "    ans=c(ans,mean(cspl[which(cspl$virTree==v),colm],na.rm=T))\n",
    "  }\n",
    "  return(ans)\n",
    "}\n",
    "lincat=function(i,colm){\n",
    "  ans=NULL\n",
    "  for(v in unique(cspl$virTree[which(cspl$Family==unique(cspl$Family)[i])])){\n",
    "    ans=c(ans,as.character(unique(cspl[which(cspl$virTree==v),colm])))\n",
    "  }\n",
    "  return(ans)\n",
    "}\n",
    "\n",
    "get_panlin=function(i,colm){\n",
    "  return(panlin[which(panlin$lineage %in% unique(cspl$virTree[which(cspl$Family==unique(cspl$Family)[i])])),colm])\n",
    "}\n",
    "\n",
    "hdiv2=cbind(family=as.character(unique(cspl[,'Family'])),gmlm=0,gmlsd=0,aanm=0,aansd=0,\n",
    "            nhost=0,hosts=0,hdiv=0,ugpnm=0,ugpnsd=0,sgpnm=0,sgpnsd=0,\n",
    "            kadisd=0,kodm=0,kodsd=0,ipsd=0,fmgpsd=0,sgt=0,\n",
    "            coregp=0,corept=0,coreppp=0,coregpp=0)\n",
    "for(i in 1:nrow(hdiv2)){\n",
    "  cat(i)\n",
    "  hdiv2[i,'gmlm']=mean(linlin(i,'Length'),na.rm=T)\n",
    "  hdiv2[i,'gmlsd']=sd(linlin(i,'Length'),na.rm=T)\n",
    "  hdiv2[i,'aanm']=mean(linlin(i,'aanum'),na.rm=T)\n",
    "  hdiv2[i,'aansd']=sd(linlin(i,'aanum'),na.rm=T) # aa number\n",
    "  hdiv2[i,'ugpnm']=mean(linlin(i,'gf.unique.sp'),na.rm=T) #\n",
    "  hdiv2[i,'ugpnsd']=sd(linlin(i,'gf.unique.sp'),na.rm=T) # unique gp\n",
    "  hdiv2[i,'sgpnm']=mean(linlin(i,'gf.shared'),na.rm=T)\n",
    "  hdiv2[i,'sgpnsd']=sd(linlin(i,'gf.shared'),na.rm=T) # shared gp\n",
    "  hdiv2[i,'kadisd']=sd(linlin(i,'kalldist'),na.rm=T)\n",
    "  hdiv2[i,'kodm']=mean(lincorediv(i),na.rm=T) # core gene seq div\n",
    "  hdiv2[i,'kodsd']=sd(lincorediv(i),na.rm=T)\n",
    "  hdiv2[i,'ipsd']=sd(linlin(i,'inpara.pt'),na.rm=T) # inparalog\n",
    "  hdiv2[i,'fmgpsd']=sd(linlin(i,'gf.unique.fm'),na.rm=T) # family unique gene\n",
    "  hdiv2[i,'sgt']=sd(linlin(i,'ptng'),na.rm=T)#singleton\n",
    "  \n",
    "  hdiv2[i,'coregp']=sd(get_panlin(i,'core.clusters'),na.rm=T)\n",
    "  hdiv2[i,'corept']=sd(get_panlin(i,'core.proteins'),na.rm=T)\n",
    "  hdiv2[i,'coreppp']=mean(get_panlin(i,'core.proteins')/(get_panlin(i,'unique.allp')+get_panlin(i,'spec.lin.allp')+\n",
    "                                get_panlin(i,'inter.lin.allp')+get_panlin(i,'inter.fm.allp')+get_panlin(i,'all.ptng')))\n",
    "  hdiv2[i,'coregpp']=mean(get_panlin(i,'core.clusters')/(get_panlin(i,'unique.all')+get_panlin(i,'spec.lin.all')+\n",
    "                                get_panlin(i,'inter.lin.all')+get_panlin(i,'inter.fm.all')))\n",
    "  \n",
    "  \n",
    "  hdiv2[i,'hdiv']=entt(table(lincat(i,'eutx'))/sum(table(lincat(i,'eutx')),na.rm=T))\n",
    "  hdiv2[i,'nhost']=length(unique(cspl[which(cspl$Family==hdiv2[i,'family']),'eutx']))\n",
    "  hdiv2[i,'hosts']=paste(unique(cspl[which(cspl$Family==hdiv2[i,'family']),'eutx']),collapse=',')\n",
    "}\n",
    "hdiv2=as.data.frame(hdiv2)\n",
    "\n",
    "\n",
    "\n",
    "pdf(paste0('hostdiv_',md,'b.pdf'),\n",
    "    width=8.268/2.1,height=11.693/4.2,paper='special')\n",
    "a=ggplot(hdiv2,aes(x=as.numeric(as.character(hdiv)),y=as.numeric(as.character(aansd))))+\n",
    "  geom_point(size=2)+theme_classic()+ geom_text(aes(label=family),vjust=1.5)+\n",
    "  labs(x=\"Supergroup-level host diversity\", y=\"SD of protein numbers\")+\n",
    "  scale_x_continuous(limits=c(-0.1,1.8)) #protein number sd\n",
    "print(a)\n",
    "\n",
    "a=ggplot(hdiv2,aes(x=as.numeric(as.character(hdiv)),y=as.numeric(as.character(sgt))))+\n",
    "  geom_point(size=2)+theme_classic()+ geom_text(aes(label=family),vjust=1.5)+\n",
    "  labs(x=\"Supergroup-level host diversity\", y=\"SD of singleton numbers\")+\n",
    "  scale_x_continuous(limits=c(-0.1,1.8)) #expand_limits(x=-0.5)\n",
    "print(a)\n",
    "\n",
    "\n",
    "a=ggplot(hdiv2,aes(x=as.numeric(as.character(hdiv)),y=as.numeric(as.character(kodm))))+\n",
    "  geom_point(size=2)+theme_classic()+ geom_text(aes(label=family),vjust=1.5)+\n",
    "  labs(x=\"Supergroup-level host diversity\", y=\"Core-gene tree divergence\")+\n",
    "  scale_x_continuous(limits=c(-0.1,1.8))+expand_limits(y=0) #divergence mean\n",
    "print(a)\n",
    "\n",
    "\n",
    "dev.off()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "id": "vGSLSb1Pgko-"
   },
   "outputs": [],
   "source": [
    "# a=ggplot(hdiv2,aes(x=as.numeric(as.character(hdiv)),y=as.numeric(as.character(coreppp))))+\n",
    "#   geom_point(size=2)+theme_classic()+ geom_text(aes(label=family),vjust=1.5)+\n",
    "#   labs(x=\"Diversity of eukaryote supergroup infected\", y=\"Core gene number proportion\")+\n",
    "#   scale_x_continuous(limits=c(-0.1,1.8)) #expand_limits(x=-0.5)\n",
    "# print(a)\n",
    "\n",
    "# a=ggplot(hdiv2,aes(x=as.numeric(as.character(hdiv)),y=as.numeric(as.character(ugpnsd))))+\n",
    "#   geom_point(size=2)+theme_classic()+ geom_text(aes(label=family),vjust=1.5)+\n",
    "#   labs(x=\"Diversity of eukaryote supergroup infected\", y=\"Unique gene cluster number variation\")+\n",
    "#   scale_x_continuous(limits=c(-0.1,1.8)) #expand_limits(x=-0.5)\n",
    "# print(a)\n",
    "# \n",
    "# a=ggplot(hdiv2,aes(x=as.numeric(as.character(hdiv)),y=as.numeric(as.character(ipsd))))+\n",
    "#   geom_point(size=2)+theme_classic()+ geom_text(aes(label=family),vjust=1.5)+\n",
    "#   labs(x=\"Diversity of eukaryote supergroup infected\", y=\"Paralog number variation\")+\n",
    "#   scale_x_continuous(limits=c(-0.1,1.8)) #expand_limits(x=-0.5)\n",
    "# print(a)\n",
    "\n",
    "\n",
    "\n",
    "# a=ggplot(hdiv2,aes(x=as.numeric(as.character(hdiv)),y=as.numeric(as.character(gmlsd))))+\n",
    "#   geom_point(size=2)+theme_classic()+ geom_text(aes(label=family),vjust=1.5)+\n",
    "#   labs(x=\"Diversity of eukaryote supergroup infected\", y=\"Genome length variation\")+\n",
    "#   scale_x_continuous(limits=c(-0.1,1.8)) #genome length sd\n",
    "# print(a)\n",
    "\n",
    "# a=ggplot(hdiv2,aes(x=as.numeric(as.character(hdiv)),y=as.numeric(as.character(coregp))))+\n",
    "#   geom_point(size=2)+theme_classic()+ geom_text(aes(label=family),vjust=1.5)+\n",
    "#   labs(x=\"Diversity of eukaryote supergroup infected\", y=\"Core gene cluster variation\")+\n",
    "#   scale_x_continuous(limits=c(-0.1,1.8)) #expand_limits(x=-0.5)\n",
    "# print(a)\n",
    "# \n",
    "# a=ggplot(hdiv2,aes(x=as.numeric(as.character(hdiv)),y=as.numeric(as.character(corept))))+\n",
    "#   geom_point(size=2)+theme_classic()+ geom_text(aes(label=family),vjust=1.5)+\n",
    "#   labs(x=\"Diversity of eukaryote supergroup infected\", y=\"Core gene number variation\")+\n",
    "#   scale_x_continuous(limits=c(-0.1,1.8)) #expand_limits(x=-0.5)\n",
    "# print(a)\n",
    "# \n",
    "# a=ggplot(hdiv2,aes(x=as.numeric(as.character(hdiv)),y=as.numeric(as.character(coregpp))))+\n",
    "#   geom_point(size=2)+theme_classic()+ geom_text(aes(label=family),vjust=1.5)+\n",
    "#   labs(x=\"Diversity of eukaryote supergroup infected\", y=\"Core gene cluster proportion\")+\n",
    "#   scale_x_continuous(limits=c(-0.1,1.8)) #expand_limits(x=-0.5)\n",
    "# print(a)\n"
   ]
  }
 ],
 "metadata": {
  "colab": {
   "authorship_tag": "ABX9TyNZ7ZiyhUKfyv/+V8c01v71",
   "name": "R_0607fig3.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 1
}
